File: minigit-3/ts/src/commands/update-ref.ts
// update-ref command - Update reference

import { findRepoRoot } from '../utils';
import { writeRef, resolveRef } from '../refs';

export function updateRef(args: string[]): void {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a git repository');
    process.exit(1);
  }

  if (args.length < 2) {
    console.error('usage: update-ref <ref> <sha>');
    process.exit(1);
  }

  const refName = args[0];
  const sha = args[1];

  // Resolve short SHA if needed
  let fullSha = sha;
  if (sha.length < 40) {
    const resolved = resolveRef(sha, repoRoot);
    if (resolved) {
      fullSha = resolved;
    }
  }

  // Validate SHA format
  if (!/^[0-9a-f]{40}$/.test(fullSha)) {
    console.error(`fatal: ${sha}: not a valid SHA1`);
    process.exit(1);
  }

  writeRef(refName, fullSha, repoRoot);
}

File: minigit-3/py/commands/update_ref.py
"""update-ref command - Update reference."""

import sys
from pathlib import Path


def run(args: list[str]) -> int:
    """Update a reference."""
    from utils import find_repo_root
    from refs import write_ref, write_head

    repo_root = find_repo_root()
    if repo_root is None:
        print("error: not a minigit repository", file=sys.stderr)
        return 1

    if len(args) < 2:
        print("error: ref and sha required", file=sys.stderr)
        return 1

    ref = args[0]
    sha = args[1]

    # Validate SHA format (should be 40 hex chars)
    if len(sha) != 40 or not all(c in '0123456789abcdef' for c in sha):
        print(f"error: invalid SHA: '{sha}'", file=sys.stderr)
        return 1

    if ref == 'HEAD':
        write_head(repo_root, sha)
    else:
        write_ref(repo_root, ref, sha)

    return 0
